src/offscreen.js

```
@@ -0,0 +1,47 @@
+chrome.runtime.onMessage.addListener(handleMessages);
+
+async function handleMessages(message) {
+  console.log('handle message called');
+  if (message.target !== 'offscreen-doc') {
+    console.log('target is bad');
+    return;
+  }
+
+  if (message.type === 'copy-data-to-clipboard') {
+    handleClipboardWrite(message.data);
+  } else {
+    console.warn(`Unexpected message type received: '${message.type}'.`);
+  }
+}
+
+// We use a <textarea> element for two main reasons:
+//  1. preserve the formatting of multiline text,
+//  2. select the node's content using this element's `.select()` method.
+const textEl = document.querySelector('#text');
+
+// Use the offscreen document's `document` interface to write a new value to the
+// system clipboard.
+//
+// At the time this demo was created (Jan 2023) the `navigator.clipboard` API
+// requires that the window is focused, but offscreen documents cannot be
+// focused. As such, we have to fall back to `document.execCommand()`.
+async function handleClipboardWrite(data) {
+  try {
+    // Error if we received the wrong kind of data.
+    if (typeof data !== 'string') {
+      throw new TypeError(
+        `Value provided must be a 'string', got '${typeof data}'.`
+      );
+    }
+
+    // `document.execCommand('copy')` works against the user's selection in a web
+    // page. As such, we must insert the string we want to copy to the web page
+    // and to select that content in the page before calling `execCommand()`.
+    textEl.value = data;
+    textEl.select();
+    document.execCommand('copy');
+  } finally {
+    // Job's done! Close the offscreen document.
+    window.close();
```

While I was testing a recent MV3 migration of one of my extensions, I discovered a race condition related to closing the offscreen doc like this immediately after calling `document.execCommand('copy');`.

The copy will occasionally fail silently (~20% failure rate iirc). That is, execCommand will return true, but the content fails to get written.

I dealt with this by adding an inactivity delay. I ran into a few other interesting quirks around managing the offscreen doc lifecycle as well.

For the curious: https://github.com/hansifer/tab-copy/blob/master/src/offscreen-actions.ts